firedrake/plot.py

```diff
@@ -0,0 +1,345 @@
+        vec_length = function.ufl_shape[0]
+    if vec_length == 1:
+        data = np.reshape(data, data.shape+(1, ))
+    return np.einsum("ijk,jl->ilk", data, elem)
+
+
+def _calculate_points(function, num_points, dimension):
+    """Calculate points in physical space of given function with given number of
+    sampling points at given dimension
+
+    :arg function: function to be sampled
+    :arg num_points: number of sampling points
+    :arg dimension: dimension of the function
+    """
+    function_space = function.function_space()
+    mesh = function_space.mesh()
+    if mesh.ufl_cell() == Cell('interval'):
+        points = np.linspace(0.0, 1.0, num=num_points,
+                             dtype=float).reshape(-1, 1)
+    elif mesh.ufl_cell() == Cell('quadrilateral'):
+        points_1d = np.linspace(0, 1.0, num=num_points,
+                                dtype=float).reshape(-1, 1)
+        points = np.array(np.meshgrid(points_1d, points_1d)).T.reshape(-1, 2)
+    elif mesh.ufl_cell() == Cell('triangle'):
+        points_1d = np.linspace(0, 1.0, num=num_points,
+                                dtype=float).reshape(-1, 1)
+        points_1d_rev = np.fliplr([points_1d]).reshape(-1)
+        iu = np.triu_indices(num_points)
+        points = np.array(np.meshgrid(points_1d, points_1d_rev)).T[iu]
+    else:
+        raise RuntimeError("Unsupported functionality")
+    y_vals = _calculate_values(function, points, dimension)
+    x_vals = _calculate_values(mesh.coordinates, points, dimension)
+    return x_vals, y_vals
+
+
+def calculate_one_dim_points(function, num_points):
+    """Calculate a set of points for plotting for a one-dimension function as a
+    numpy array
+
+    :arg function: 1D function for plotting
+    :arg num_points: Number of points per element
+    """
+    x_vals, y_vals = _calculate_points(function, num_points, 1)
+    x_vals = x_vals.reshape(-1)
+    y_vals = y_vals.reshape(-1)
+    order = np.argsort(x_vals)
+    x_vals = x_vals[order]
+    y_vals = y_vals[order]
+    return np.array([x_vals, y_vals])
+
+
+def two_dimension_triangle_Z(function, num_sample_points):
+    """Calculate the triangulation and function values for a given 2D function
+
+    :arg function: 2D function
+    :arg num_sample_points: Number of sampling points
+    """
+
+    try:
+        from matplotlib.tri import Triangulation, UniformTriRefiner
+    except ImportError:
+        raise RuntimeError("Matplotlib not importable, is it installed?")
+    if function.function_space().mesh().cell() == Cell('triangle'):
+        x = np.array([0, 0, 1])
+        y = np.array([0, 1, 0])
+    elif function.function_space().mesh().cell() == Cell('quadrilateral'):
+        x = np.array([0, 0, 1, 1])
+        y = np.array([0, 1, 0, 1])
+    else:
+        raise RuntimeError("Unsupported Functionality")
+    base_tri = Triangulation(x, y)
+    refiner = UniformTriRefiner(base_tri)
+    tri = refiner.refine_triangulation(False, num_sample_points)
```

Comment:
This produces `4**num_sample_points` triangles per input triangle.  Which is probably not what you wanted!